stomp_plugins/src/noise_generators/cartesian_constraints_sampling.cpp

```diff
@@ -0,0 +1,249 @@
+    for(auto i = 0u; i < stddev_param.size(); i++)
+    {
+      stddev_[i] = static_cast<double>(stddev_param[i]);
+    }
+  }
+  catch(XmlRpc::XmlRpcException& e)
+  {
+    ROS_ERROR("%s failed to load parameters",getName().c_str());
+    return false;
+  }
+
+  return true;
+}
+
+bool CartesianConstraintsSampling::setMotionPlanRequest(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                 const moveit_msgs::MotionPlanRequest &req,
+                 const stomp_core::StompConfiguration &config,
+                 moveit_msgs::MoveItErrorCodes& error_code)
+{
+  bool succeed = setupNoiseGeneration(planning_scene,req,config,error_code) &&
+          setupRobotState(planning_scene,req,config,error_code);
+  return succeed;
+}
+
+bool CartesianConstraintsSampling::setupNoiseGeneration(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                                               const moveit_msgs::MotionPlanRequest &req,
+                                               const stomp_core::StompConfiguration &config,
+                                               moveit_msgs::MoveItErrorCodes& error_code)
+{
+  using namespace Eigen;
+
+  // convenience lambda function to fill matrix
+  auto fill_diagonal = [](Eigen::MatrixXd& m,double coeff,int diag_index)
+  {
+    std::size_t size = m.rows() - std::abs(diag_index);
+    m.diagonal(diag_index) = VectorXd::Constant(size,coeff);
+  };
+
+  // creating finite difference acceleration matrix
+  std::size_t num_timesteps = config.num_timesteps;
+  Eigen::MatrixXd A = MatrixXd::Zero(num_timesteps,num_timesteps);
+  int num_elements = (int((ACC_MATRIX_DIAGONAL_INDICES.size() -1)/2.0) + 1)* num_timesteps ;
+  for(auto i = 0u; i < ACC_MATRIX_DIAGONAL_INDICES.size() ; i++)
+  {
+    fill_diagonal(A,ACC_MATRIX_DIAGONAL_VALUES[i],ACC_MATRIX_DIAGONAL_INDICES[i]);
+  }
+
+  // create and scale covariance matrix
+  Eigen::MatrixXd covariance = MatrixXd::Identity(num_timesteps,num_timesteps);
+  covariance = A.transpose() * A;
+  covariance = covariance.fullPivLu().inverse();
+  double max_val = covariance.array().abs().matrix().maxCoeff();
+  covariance /= max_val;
+
+  // preallocating noise data
+  raw_noise_ = Eigen::VectorXd::Zero(CARTESIAN_DOF_SIZE);
+
+  error_code.val = error_code.SUCCESS;
+
+  return true;
+}
+
+bool CartesianConstraintsSampling::setupRobotState(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                                               const moveit_msgs::MotionPlanRequest &req,
+                                               const stomp_core::StompConfiguration &config,
+                                               moveit_msgs::MoveItErrorCodes& error_code)
+{
+  using namespace moveit::core;
+  using namespace utils::kinematics;
+
+  // robot state
+  const JointModelGroup* joint_group = robot_model_->getJointModelGroup(group_);
+  tool_link_ = joint_group->getLinkModelNames().back();
+  state_.reset(new RobotState(robot_model_));
```

Comment:
in constructor and make_shared / make_unique

Reply:
Each noise generator initial robot_state in virtual method setupRobotState()